File: src/main/java/me/mrafonso/runway/listeners/SystemChatListener.java
package me.mrafonso.runway.listeners;

import com.github.retrooper.packetevents.event.simple.PacketPlaySendEvent;
import com.github.retrooper.packetevents.protocol.packettype.PacketType;
import com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerSystemChatMessage;
import me.mrafonso.runway.config.ConfigManager;
import me.mrafonso.runway.util.ProcessHandler;
import net.kyori.adventure.text.minimessage.MiniMessage;
import org.bukkit.entity.Player;

public class SystemChatListener extends AbstractListener {
    private final MiniMessage mm = MiniMessage.miniMessage();

    public SystemChatListener(ProcessHandler processHandler, ConfigManager configManager) {
        super(processHandler, configManager);
    }

    @Override
    public void onPacketPlaySend(PacketPlaySendEvent e) {
        if (!config.getOrDefault("listeners.system-messages", true) ||
            e.getPacketType() != PacketType.Play.Server.SYSTEM_CHAT_MESSAGE) return;

        Player player = e.getPlayer();
        WrapperPlayServerSystemChatMessage packet = new WrapperPlayServerSystemChatMessage(e);
        String message = mm.serialize(packet.getMessage());

        if (message.startsWith("<lang:multiplayer.message_not_delivered:")) {
            e.setCancelled(true);
            return;
        }

        if (message.contains("[actionbar]")) {
            message = message.replace("[actionbar]", "");
            e.setCancelled(true);
            player.sendActionBar(handler.processComponent(message, player));
        } else {
            packet.setMessage(handler.processComponent(message, player));
        }
    }
}

File: src/main/kotlin/me/mrafonso/runway/listener/SystemChatListener.kt
package me.mrafonso.runway.listeners

import com.github.retrooper.packetevents.event.simple.PacketPlaySendEvent
import com.github.retrooper.packetevents.protocol.packettype.PacketType
import com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerSystemChatMessage
import me.mrafonso.runway.config.Settings
import me.mrafonso.runway.config.ConfigHandler
import me.mrafonso.runway.processing.ProcessHandler
import me.mrafonso.runway.listener.AbstractPacketListener
import org.bukkit.entity.Player
import java.awt.TextComponent

class SystemChatListener(processHandler: ProcessHandler, configHandler: ConfigHandler) :
    AbstractPacketListener(processHandler, configHandler) {

    override fun onPacketPlaySend(e: PacketPlaySendEvent) {
        super.onPacketPlaySend(e)
        if (e.packetType != PacketType.Play.Server.SYSTEM_CHAT_MESSAGE) return

        val settings = configHandler.get<Settings>()
        if (!settings.listeners.systemMessages) return

        val player = e.getPlayer<Player?>()
        val packet = WrapperPlayServerSystemChatMessage(e)

        val text = mm.serialize(packet.message)

        if (text.startsWith("<lang:multiplayer.message_not_delivered:")) {
            e.isCancelled = true
            return
        }

        packet.message = handler.processComponent(text, player) ?: return

        if (settings.prefix.required) text.drop(settings.prefix.value.length)
        if (text.contains("\\<silent>")) {
            e.isCancelled = true
        }
    }
}